lib/LedgerEventStorage.js

```diff
@@ -545,4 +521,80 @@ class LedgerEventStorage {
   }
 }
 
+function _createUpdatePatch({patch, idempotent, id}) {
+  if(!Array.isArray(patch)) {
+    throw new TypeError('patch must be an array');
+  }
+
+  const setObject = {};
+  const unsetObject = {};
+  const pushFields = {};
+  const pullFields = {};
+
+  for(const operation of patch) {
+    if(idempotent && !VALID_IDEMPOTENT_OPERATIONS.has(operation.op)) {
+      throw new BedrockError(
+        'Only idempotent operations are allowed.',
+        'NotAllowedError',
+        {operation, validOperations: Array.from(VALID_IDEMPOTENT_OPERATIONS)}
+      );
+    }
+
+    // ensure that only meta fields are modified
+    const opLength = Object.keys(operation.changes).length;
+    if(opLength !== 1 ||
+      (opLength === 1 && operation.changes.meta === undefined)) {
+      throw new BedrockError(
+        'Only event meta can be updated.',
+        'NotAllowedError', {operation});
+    }
+
+    // process set, unset, add, and remove operations
+    if(operation.op === 'set') {
+      _.merge(setObject, operation.changes);
+    } else if(operation.op === 'unset') {
+      _.merge(unsetObject, operation.changes);
+    } else if(operation.op === 'add') {
+      const arrayUpdate = database.buildUpdate(operation.changes);
+      const field = Object.keys(arrayUpdate)[0];
+      if(field in pushFields) {
+        pushFields[field].$each.push(arrayUpdate[field]);
+      } else {
+        pushFields[field] = {$each: [arrayUpdate[field]]};
+      }
+    } else if(operation.op === 'remove') {
+      const arrayUpdate = database.buildUpdate(operation.changes);
+      const field = Object.keys(arrayUpdate)[0];
+      if(field in pullFields) {
+        pullFields[field].push(arrayUpdate[field]);
+      } else {
+        pullFields[field] = [arrayUpdate[field]];
+      }
+    }
+  }
+
+  // build the update object for MongoDB
+  const update = {$set: {}};
+  const setFields = database.buildUpdate(setObject);
+  const unsetFields = database.buildUpdate(unsetObject);
+
+  if(Object.keys(setFields).length > 0) {
+    update.$set = setFields;
+  }
+  if(Object.keys(unsetFields).length > 0) {
+    update.$unset = unsetFields;
+  }
+  if(Object.keys(pushFields).length > 0) {
+    update.$addToSet = pushFields;
+  }
+  if(Object.keys(pullFields).length > 0) {
+    update.$pullAll = pullFields;
+  }
+
+  // add update id
+  update.$set['meta.uid'] = Buffer.from(id);
```

Comment:
Let's see if we can use:

https://docs.mongodb.com/manual/reference/method/BulkWriteResult/#BulkWriteResult.writeErrors.op

Instead of needing an extra field here.